twine/commands/upload.py

```diff
@@ -30,6 +34,26 @@
 def skip_upload(
     response: requests.Response, skip_existing: bool, package: package_file.PackageFile
 ) -> bool:
+    """Skip uploading a package.
+
+    Return Boolean type according to the status code that responded by the repository or
+    the argument passed by the user when trying to upload the package(s).
+
+    If ``skip_existing`` is set to ``True``, then return ``False``.
+    If status code 400, 403, 409 is responded by the repository, return ``True``.
+
+    :param requests.Response response:
+        Get the response from the repository.
+    :param bool skip_existing:
+        Specify whether twine should continue uploading files if one
+        of them already exists. This primarily supports PyPI. Other
+        package indexes may not be supported.
+    :param package_file.PackageFile package:
+        Get the package files.
+
+    :return bool:
+        Determine whether we should skip uploading the package.
+    """
```

Comment:
The `:param` and `:return` directives feel redundant to me, given the type annotations and preceding prose. Twine only uses these directives in the [`Settings class`](https://github.com/pypa/twine/blob/b4a196ee452363662adf123b4546754271b23e57/twine/settings.py#L67) and the [`get_userpass_value` function](https://github.com/pypa/twine/blob/b4a196ee452363662adf123b4546754271b23e57/twine/utils.py#L207)

Futhermore, the primary reference that I could find for the [Sphinx docstring style](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html) uses an explicit `:type` directive, instead of inlining the type in `:param`. The only "official" documentation of the inline style that I could find is buried in the [Sphinx "Domains" documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists).

To reduce maintenance overhead and improve readability, I'd prefer to remove the Sphinx directives in favor of explanatory prose as needed. Alternatively, we could reduce the prose in favor of the directives, in which case I'd remove the type declarations.

@sigmavirus24 Do you feel strongly one way or another?

Reply:
@sigmavirus24 Any preferences re: Sphinx directives? Are there any other type-annotated projects that you'd recommend as an example of documenting arguments?

Reply:
If we were to have a separate section of documentation for contributors and maintainers, this would be useful. Flake8 has that documentation and uses this style of doc string for internal only classes (example: [`Application`](https://github.com/PyCQA/flake8/blob/master/src/flake8/main/application.py))

Reply:
Okay. I can also see how directives could give useful structure to contributors writing docstrings, instead of arbitrary prose.

@meowmeowmeowcat With that in mind, I'd like to take a pass at editing what you've put together, and experimenting with a few things, before handing it back to you.